### PR TITLE
Fix reading heap overflow in a test case

### PR DESCRIPTION
**Pull request by InfoHunter**
This is caught by AddressSanitizer. Reading goes beyond the boundary...

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

File: test/sslapitest.c [hunk 1]
```diff
@@ -1970,7 +1970,7 @@ static int alpn_select_cb (SSL *ssl, const unsigned char **out, unsigned char *o
     unsigned int i, protlen = 0;
     const unsigned char *prot;
 
-    for (i = 0, prot = in; i < inlen; i += protlen, prot += protlen) {
+    for (i = 0, prot = in; i < inlen - 1; i += protlen, prot += protlen) {
         protlen = *(prot++);
```
Comment (mattcaswell) [inline test/sslapitest.c]: This fix doesn't look right to me. Shouldn't there be an `i++` at the same time as we do the `prot++` here?

Comment (InfoHunter) [reply test/sslapitest.c]: sounds right, fixed...

**Comment by InfoHunter** [comment]
iterator `i` seems not necessary, so how about this?

File: test/sslapitest.c [hunk 2]
```diff
@@ -1967,12 +1967,12 @@ static const char *servalpn;
 static int alpn_select_cb (SSL *ssl, const unsigned char **out, unsigned char *outlen,
                     const unsigned char *in, unsigned int inlen, void *arg)
```
Comment (dot-asm) [inline test/sslapitest.c]: While you are at it, fix width and indentation here.

Comment (InfoHunter) [reply test/sslapitest.c]: Sure, fixed

File: test/sslapitest.c [hunk 3]
```diff
@@ -1967,12 +1967,12 @@ static const char *servalpn;
...
 
-    for (i = 0, prot = in; i < inlen; i += protlen, prot += protlen) {
-        protlen = *(prot++);
-        if (inlen - i < protlen)
+    for (prot = in; prot < in + inlen; prot += protlen) {
+        protlen = *prot++;
+        if (in + inlen - prot < protlen)
             return SSL_TLSEXT_ERR_NOACK;
 
         if (protlen == strlen(servalpn)
```
Comment (dot-asm) [inline test/sslapitest.c]: Why isn't this strlen for "goodalpn"? I mean since following memcmp is looking for "goodalpn" then one can as well compare to length of "goodalpn"...

Comment (InfoHunter) [reply test/sslapitest.c]: I didn't go through all cases in this file so I am not sure if this is intended here. One reason of not strlen for `"goodalpn"` is that the global `servalpn` would be changed to control the result of this `alpn_select_cb` callback, so it seems the `memcmp` should be changed to compare against `servalpn` instead of `"goodalpn"`...

Comment (mattcaswell) [reply test/sslapitest.c]: Yes, I think that was the intention of this code

Comment (dot-asm) [reply test/sslapitest.c]: > Yes, I think that was the intention of this code

What was? To memcmp against servalpn and not "goodalpn"? Well, that would make sense too. Basically I failed to see the reason for why input to strlen and memcmp would be *different*. I assumed that intension was to catch "goodalpn", but it might as well be to catch servalpn whatever it might happen to be at the given moment...

Comment (mattcaswell) [reply test/sslapitest.c]: I meant the intention was to memcmp against servalpn.

Comment (dot-asm) [reply test/sslapitest.c]: So it's kind of a bug. Do we fix it here or in separate request? I'd say it's not inappropriate to fix it right away, but separate commit might be appropriate...

Comment (InfoHunter) [reply test/sslapitest.c]: The fix could be very straightforward, so this should be fixed right away - in a separate commit in this PR...

**Comment by InfoHunter** [comment]
New commit added and this should *not* be squashed...

**Comment by richsalz** [comment]
Okay, what you should do is "git rebase -i master" and squash the first four.  Then force-push.

**Comment by dot-asm** [comment]
Red-cross is from Travis is unrelated.

**Comment by dot-asm** [comment]
Well, since @richsalz wants to roll it own way, then I'll keep away...

**Comment by richsalz** [comment]
If you have a better suggestion, go for it.  I was just telling @InfoHunter how to do what he wants done.

**Comment by dot-asm** [comment]
Well, I roll by #3963, where it's suggested that squashing should be committer's responsibility, not contributor's, past initial approval...

**Comment by richsalz** [comment]
but this has special requirements and the committer, especially if it's me, might skip/forget them.

**Comment by dot-asm** [comment]
So I'll keep away :-)

**Comment by InfoHunter** [comment]
Yes, it's done now :-)

**Comment by mattcaswell** [comment]
Still needs second review, @openssl/committers.

File: test/sslapitest.c [hunk 4]
```diff
@@ -1964,19 +1964,20 @@ static int hostname_cb(SSL *s, int *al, void *arg)
...
+static int alpn_select_cb(SSL *ssl, const unsigned char **out,
+                          unsigned char *outlen, const unsigned char *in,
+                          unsigned int inlen, void *arg)
 {
-    unsigned int i, protlen = 0;
+    unsigned int protlen = 0;
     const unsigned char *prot;
 
-    for (i = 0, prot = in; i < inlen; i += protlen, prot += protlen) {
-        protlen = *(prot++);
```
Comment (bernd-edlinger) [inline test/sslapitest.c]: I wonder if adding a simple "i++" here would be better fix.
I for one find it hard to follow the pointer arithmetics in "in + inlen - prot < protlen".

Comment (InfoHunter) [reply test/sslapitest.c]: So if it's like this: `end = in + inlen` and then `end - prot < protlen`. This should be more comprehensible. The unnecessary `i` is just the reason causing this bug...

**Comment by InfoHunter** [comment]
Ping @openssl ? seems this could be merged now...

**Comment by mattcaswell** [comment]
Pushed. Thanks.
